src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,450 @@
+		else {
+			c &= 0xFF >> (len + 1);
+			cp = c << ((len - 1) * 6);
+			while(--len)
+			{
+				if(++it == sv.end()) return r; // string is truncated
+				c = *it;
+				c &= 0x7F;
+				cp |= c << ((len - 1) * 6);
+			}
+		}
+
+		if(cp <= 0xFF) r.push_back(static_cast<unsigned char>(cp));
+		++it;
+	}
+	return r;
+}
+
+websocket_tracker_connection::websocket_tracker_connection(io_context& ios
+		, tracker_manager& man
+		, tracker_request const& req
+		, std::weak_ptr<request_callback> cb)
+	: tracker_connection(man, req, ios, cb)
+	  , m_io_context(ios)
+	  , m_websocket(std::make_shared<aux::websocket_stream>(m_io_context, m_man.host_resolver(), req.ssl_ctx))
+{
+	aux::session_settings const& settings = m_man.settings();
+
+	// in anonymous mode we omit the user agent to mitigate fingerprinting of
+    // the client. Private torrents is an exception because some private
+    // trackers may require the user agent
+    std::string const user_agent = settings.get_bool(settings_pack::anonymous_mode)
+			&& !tracker_req().private_torrent ? "" : settings.get_str(settings_pack::user_agent);
+	m_websocket->set_user_agent(user_agent);
+
+	queue_request(req, cb);
+}
+
+websocket_tracker_connection::~websocket_tracker_connection()
+{
+	close();
+}
+
+void websocket_tracker_connection::start()
+{
+	if(m_websocket->is_open() || m_websocket->is_connecting())
+		return;
+
+    std::shared_ptr<request_callback> cb = requester();
+    if (cb) cb->debug_log("*** WEBSOCKET_TRACKER_CONNECTING [ url: %s ]", tracker_req().url.c_str());
+
+	std::shared_ptr<websocket_tracker_connection> self(shared_from_this());
+	m_websocket->async_connect(tracker_req().url, std::bind(&websocket_tracker_connection::on_connect, self, _1));
+}
+
+void websocket_tracker_connection::close()
+{
+	if(m_websocket->is_open() || m_websocket->is_connecting())
+	{
+		m_websocket->close();
+	}
+}
+
+void websocket_tracker_connection::queue_request(tracker_request req, std::weak_ptr<request_callback> cb)
+{
+	m_pending.emplace(tracker_message{std::move(req)}, cb);
+	if(m_websocket->is_open()) send_pending();
+}
+
+void websocket_tracker_connection::queue_answer(tracker_answer ans)
+{
+	m_pending.emplace(tracker_message{std::move(ans)}, std::weak_ptr<request_callback>{});
+	if(m_websocket->is_open()) send_pending();
+}
+
+void websocket_tracker_connection::send_pending()
+{
+	if(m_sending || m_pending.empty()) return;
+
+	m_sending = true;
+
+	tracker_message msg;
+    std::weak_ptr<request_callback> cb;
+    std::tie(msg, cb) = std::move(m_pending.front());
+	m_pending.pop();
+
+	// Update requester
+	if(cb.lock()) m_requester = cb;
+
+	boost::apply_visitor([&](auto const& m)
+		{
+			do_send(m);
+		}
+		, msg
+	);
+}
+
+void websocket_tracker_connection::do_send(tracker_request const& req)
+{
+	// Update request
+	m_req = req;
+
+	json payload;
+	payload["action"] = "announce";
+	payload["info_hash"] = from_latin1({req.info_hash.data(), std::size_t(req.info_hash.size())});
```

Comment:
since `from_latin1()` takes a span, you should be able to just pass in `req.info_hash` to it. Same thing with all other calls to `from_latin1()`, you shouldn't have to call `.data()` and `.size()` manually, `span` can be constructed from anything that has a `.data()` and `.size()` member function (and returns the appropriate type)

Reply:
Ah, that's great! Sorry I didn't get this at first.